README
DBIx-ORM-Declarative version 0.19
======================

DBIx::ORM::Declarative - provides object-oriented access to a database.

DBIx::ORM::Declarative encapsulates the creation of database table classes at
compile time.  You declare the properties of table and row classes with
data structures passed to the DBIx::ORM::Declarative module when the module is
used.  You can also add table classes on the fly.  Best of all, you don't
need to know a thing about SQL to create or use the classes.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

   DBI >= 1.30
   Scalar::Util

COPYRIGHT AND LICENCE

Copyright (C) 2006, 2007 by Jim Schneider

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.6.0 or,
at your option, any later version of Perl 5 you may have available.

NOTE

_*PLEASE*_ see README.MySQL for information on how to set up and run
the full test suite (such as it is).  You will need a MySQL database
set up to run them - if you want to convert the tests to another database
please send me your patches :-D


Changes
Revision history for Perl extension DBIx::ORM::Declarative.

0.01  Thu Feb  2 19:36:25 2006
	- original version; created by h2xs 1.23 with options
		-AOX -b 5.6.0 -n DB::Access --use-old-tests --skip-exporter

0.03  Thu Mar  9 21:17:20 2006
    - Added support for joins and sequences

0.04  Sat Mar 11 18:12:47 2006
    - Renamed to DBIx::Class::Simple
    - Added support for using scalar refs to signify column names in
      search() and delete() methods.

0.05  Sun Mar 12 12:21:11 2006
    - Added support for group_by and order_by clauses
    - Added explicit aliases to generated SQL SELECT statements
    - Added support for limits on searches.

0.06  Tue Mar 14 19:23:19 2006
    - Added size() table method.
    - Moved to DBIx::ORM::Declarative name.  This is the final name.

0.07  Mon Apr 10 18:40:01 2006
    - Fixed a typo in __create_where so that multiple clause construction
      actually works.

0.08  Mon Apr 10 20:49:01 2006
    - Fixed another typo.

0.09  Sun Jul  2 10:06:17 2006
    - Fixed 'Use of uninitialized value in string ne' warning message.
    - Fixed missing name in automagic columns created from unique declarations.

0.10  Thu Sep 14 18:37:06 2006
    - Removed perl 5.6 specific code.
    - Changed warning suppression code to use "local ($SIG{__WARN__})"
    - Added create_only method, which creates one or more rows without
      validating the input or returning row objects.
    - Added WHERE clause handling to the size method.

0.11  Fri Sep 15 14:26:33 2006
    - Fixed bug in WHERE clause handing in size() method.

0.12  Sun Nov  5 20:32:47 2006
    - Fixed missed case in apply_method; now, coderefs also work as methods.

0.13  Mon Nov  6 20:53:10 2006
    - Added subselect processing for 'IN'/'NOT IN' WHERE clauses.

0.14  Fri Jun  1 22:37:01 2007
    Good grief, what *didn't* change?  There's more than 1500 new lines!
    - Reorganized and annotated the code extensively
    - Added the beginnings of a real test suite
    - Added methods to handle creating lots of rows at once
    - Added classes and methods to better deal with joins; not perfect, but
      it's better than the earlier join objects.
    - Added Module::Build-based Build.PL file.  The Makefile.PL file is
      deprecated, and may be removed in the future.

0.15  Tue Jul  3 20:16:29 2007
    - Changed table/join creation to better support tables with column names
      that match the table name.

0.16  Wed Jul 25 21:29:55 2007
    - Fixed table aliases bug introduced with 0.15; aliases now update the
      table_methods hash
    - Added implicit subclassing - now, when you do this:

    package Foo;
    use DBIx::ORM::Declarative
    (
        {
            schema => 'fooschema',
            # The rest of the declaration...
        },
    ) ;

    you can just do this:

    my $db = new Foo handle => $dbh;

    and $db is a schema object.  Note that if you have multiple schemas,
    Foo will be added as a subclass to all of them.

0.17  Sat Aug 18 10:22:26 2007
    - Fixed schema class magic to work better with classes that provide
      an AUTOLOAD function

0.18  Mon Aug 20 18:57:21 2007
    - Added support for extensive joins; now you can do joins across sets of
      tables where no single table has references to all of the others.

0.19  Sat Sep  1 09:30:06 2007
    - Added functionality to automatically connect to a database.

0.20  Wed Sep  5 08:50:43 2007
    - Expanded bulk_create to account for all of the unique keys - not just the
      primary keys.

0.21  Wed Sep  5 09:17:07 2007
    - Fixed a couple of typos


lib/DBIx/ORM/Declarative/Table.pm
package DBIx::ORM::Declarative::Table;

use strict;
use Carp;
use Scalar::Util qw(reftype);

use vars qw(@ISA);
@ISA = qw(DBIx::ORM::Declarative::Schema);

# The default where clause prefix:
sub _where_prefix { ''; }

# We're not a joing
sub _isjoin { 0; }

# Handle a rollback operation
sub __do_rollback
{
    my ($self, @ops) = @_;
    my $handle = $self->handle;
    carp "Can't roll back: no database" and return unless $handle;

    # Do the passed-in undo operations
    $handle->do($_) foreach @ops;

    # Turn off warnings
    local($SIG{__WARN__}) = $self->w__noop;
    $handle->rollback;
}

# The constraints
# Check for a number
sub isnumber
{
    my ($self, $value) = @_;
    return unless defined $value;
    $value =~ /^(?:\.\d+)|(?:\d+(?:\.\d+)?)$/;
}

# Check for any defined string
sub isstring
{
    my ($self, $value) = @_;
    defined $value and length $value;
}

# Check for a number, or nothing
sub isnullablenumber
{
    my ($self, $value) = @_;
    return defined $value?$self->isnumber($value):1;
}

# Always passes
sub isnullablestring { 1; }

# How to handle operators in a search operation
use constant criteriamap =>
{
    eq => [ '=', 1 ],
    ne => [ '!=', 1 ],
    gt => [ '>', 1 ],
    lt => [ '<', 1 ],
    ge => [ '>=', 1 ],
    le => [ '<=', 1 ],
    isnull => [ 'IS NULL', 0 ],
    notnull => [ 'IS NOT NULL', 0 ],
    in => [ undef, 'IN' ],
    notin => [ undef, 'NOT IN' ],
    like => [ 'LIKE', 1 ],
    notlike => [ 'NOT LIKE', 1 ],
} ;

# How many parameters are taken in limit and order clauses
use constant critexceptions =>
{
    'limit by' => 2,
    'order by' => 1,
};

# Look for a 'limit by' clause
sub __find_limit
{
    my ($self, @critera) = @_;
    my ($offset, $count) = $self->__do_special_purpose('limit by', @critera);
    return unless defined $count;
    my $lc = $self->_limit_clause;
    $lc =~ s/%offset%/$offset/g;
    $lc =~ s/%count%/$count/g;
    $lc;
}

# Handle an "order by" clause we've found
sub __find_orderby
{
    my ($self, @critera) = @_;
    my ($colref) = $self->__do_special_purpose('order by', @critera);
    return unless ref $colref;
    'ORDER BY ' . join(',', @$colref);
}

# Pull out a special-purpose clause
sub __do_special_purpose
{
    my ($self, $clause, @criteria) = @_;

    # Go through each search critereon
    for my $crit (@criteria)
    {
        # Look at each element of the critereon
        my @subcrit = @$crit;
        while(@subcrit)
        {
            my $col = shift @subcrit;

            # Did we find the clause?
            if($col eq $clause)
            {
                return @subcrit;
            }

            # Remove any parameters for a special-purpose clause
            my $cnt = $self->critexceptions->{$col} || 0;
            splice @subcrit, 0, $cnt;
            
            # Eat any following parameters
            splice @subcrit, 0, $self->criteriamap->{$col}[1]
                if $self->criteriamap->{$col};
        }
    }
    return;
}

# Create a where clause for use in searching
sub __create_where
{
    my ($self, @criteria) = @_;

    # Keep track of where clause components and the data bound to it
    my @clauses;
    my @binds;

    # The perl to SQL name map
    my %map = $self->_column_map;

    # Iterate over each critereon
    for my $crit (@criteria)
    {
        # The components that make up this chunk of the where clause
        my @sclause = ();

        my @subcrit = @$crit;
        while(@subcrit)
        {
            my $col = shift @subcrit;
            my $cnt = $self->critexceptions->{$col};

            # Is this a special-purpose clause?
            if($cnt)
            {
                splice @subcrit, 0, $cnt;
                next;
            }
            
            # Get the operator for this column
            my $op = shift @subcrit;
            my $test;

            # Do we actually have a column with this name?
            carp "No such column $col" and return unless $map{$col};

            # Is it a regular operator?
            if(defined $self->criteriamap->{$op}[0])
            {
                # Does it take a parameter?
                if($self->criteriamap->{$op}[1])
                {
                    $test = "$map{$col} " . $self->criteriamap->{$op}[0] . ' ';
                    my $val = shift @subcrit;

                    # Handle literal expressions
                    if(ref $val)
                    {
                        $test .= $$val;
                    }

                    # Or a to-be-quoted value
                    else
                    {
                        $test .= '?';
                        push @binds, $val;
                    }
                }
                else    # No parameter
                {
                    $test = "$col " . $self->criteriamap->{$op}[0];
                }
            }
            else    # IN/NOT IN
            {
                $test = "$map{$col} " . $self->criteriamap->{$op}[1] . ' (';
                my $val = shift @subcrit;

                if('SCALAR' eq reftype $val)
                {
                    # It's a subselect, or other literalized expression
                    $test .= $$val;
                }
                elsif('ARRAY' eq reftype $val)
                {
                    # It's an array of values
                    $test .= join(',', ('?')x@$val);
                    push @binds, @$val;
                }
                else
                {
                    # Treat it like a single-element list
                    $test .= '?';
                    push @binds, $val;
                }

                $test .= ')';
            }
            push @sclause, $test;
        }

        # Stick the pieces together
        push @clauses, join(' AND ', @sclause) if @sclause;
    }

    # join the subclauses together
    # Wrap them in parens if there are more than one
    @clauses = map { "($_)" } @clauses if @clauses > 1;

    # Join them together
    my $where = join(' OR ', @clauses) || '';
    
    # Add any required prefix
    my $where_pre = $self->_where_prefix;
    if($where_pre)
    {
        if($where)
        {
            $where = "($where_pre) AND ($where)";
        }
    }
    return ($where, @binds);
}

# Creates one or more items
# Does not return row objects
# Does not validate the input
# Returns an array of undef or 1 values (depending on reported success)
sub create_only
{
    my ($self, @data) = @_;

    # parameter checking
    my $handle = $self->handle;
    carp "can't create without a database handle" and return unless $handle;
    carp "can't create a row in a JOIN" and return if $self->_join_clause;

    my $table = $self->_sql_name;
    my @cols = map { $_->{name}; } $self->_columns;
    my %name2sql = $self->_column_map;
    my @rv = ();

    # We really don't want any warnings...
    local ($SIG{__WARN__}) = $self->w__noop;
    for my $row (@data)
    {
        # Execute a statement per data item
        my @use_cols = grep { exists $row->{$_}; } @cols;
        my $sql = "INSERT INTO $table (" . join(',', @name2sql{@use_cols})
            . ') VALUES (' . join(',', ('?') x @use_cols) . ')';

        # Get a statement handle
        my $sth = $handle->prepare_cached($sql);
        push @rv, undef and next unless $sth;

        # Execute and save the result
        my $rc = $sth->execute(@{$row}{@use_cols});
        push @rv, $rc?1:undef;
    }
    $handle->commit;
    return @rv;
}

# Creates multiple rows, returns the number of rows created (or
# whatever the handle object says is the number of rows)
sub bulk_create
{
    # $cols_ref is an array of column (alias) names
    my ($self, $cols_ref, @data) = @_;

    my $handle = $self->handle;
    carp "can't create without a database handle" and return unless $handle;
    carp "can't create a row in a JOIN" and return if $self->_join_clause;

    my $table = $self->table;
    carp "can't create a row without a table" and return unless $table;
    my @cols = map { $_->{name}; } $self->_columns;
    my %name2sql = $self->_column_map;
    
    my @col_unk = grep { not exists $name2sql{$_} } @$cols_ref;

    warn "Unknown columns '" . join("', '", @col_unk) . "'" and return
        if @col_unk;

    # Map unique keys to avoid duplicates
    my @uniqs_map;
    for my $un ($self->_unique_keys)
    {
        my $h = { };
        for my $i (0..$#$cols_ref)
        {
            $h->{$name2sql{$cols_ref->[$i]}} = $i;
        }
        push @uniqs_map, $h if %$h;
    }

    my $sql = "INSERT INTO $table (" . join(',', @name2sql{@$cols_ref}) . ') ';

    # We build the complete insert statement from a bunch of select statements
    # pasted together with UNION ALL
    # To avoid errors, we use another select to make sure the row is unique
    my @selects;
    for my $d (@data)
    {
        my $sel = 'SELECT ' . join(',', map { $handle->quote($_) } @$d)
            . ' FROM DUAL';
        if(@uniqs_map)
        {
            my @wherefrag = ();
            for my $un (@uniqs_map)
            {
                my @wk = map { "$_=" . $handle->quote($d->[$un->{$_}]) }
                    keys %$un;
                push @wherefrag, join(' AND ', @wk);
            }
            $sel .= ' WHERE NOT EXISTS (SELECT 1 FROM $table WHERE (' .
                join(') OR (', @wherefrag) . '))';
        }
        push @selects, $sel;
    }

    $sql .= join(' UNION ALL ', @selects);
    my $res = $handle->do($sql);

    # We don't need warnings about commit being ineffective
    local ($SIG{__WARN__}) = $self->w__noop;
    $handle->commit;
    return $res;
}

# Check parameters against the declared constraints and create
# a row in a table, returning the corresponding row object.
sub create
{
    my ($self, %params) = @_;
    my $handle = $self->handle;
    carp "can't create without a database handle" and return unless $handle;
    carp "can't create a row in a JOIN" and return if $self->_join_clause;

    # Get the data we'd need to do the create
    my ($flag, $keys, $values, $npk, @binds) =
        $self->__check_constraints($self, %params);
    return unless $flag;

    # Generate the SQL command
    my $sql = 'INSERT INTO ' . $self->_table . " ($keys) VALUES ($values)";

    # Run the command
    unshift @binds, undef if @binds;    # Avoid DBI breakage
    my $res = $handle->do($sql, @binds);
    carp "Database error: ", $handle->errstr and return unless $res;

    # Get return information
    my @res;

    # Handle the case where the primary key is null
    if($npk)
    {
        my $np = $self->_select_null_primary;
        if($np)
        {
            my $data = $handle->selectall_arrayref($np);
            if(not $data or not defined $data->[0][0])
            {
                carp "Database error: ", $handle->errstr;
                $self->__do_rollback;
                return;
            }
            @res = $self->search([
                map {($_, 'eq', $data->[0][0])} $self->_primary_key]);
        }
    }

    # Handle defined unique keys
    elsif($self->_unique_keys)
    {
        my ($un) = $self->_unique_keys;
        my @pk = @$un;
        
        # We search by the first unique key we find
        @res = $self->search([
            map {($_, (defined $params{$_}?(eq => $params{$_}):('isnull'))) }
            @pk ]);
    }

    # No unique key - do it based on everything we've got in params
    else
    {
        # This does a search on all passed-in parameters
        @res = $self->search([
            map {($_, (defined $params{$_}?(eq => $params{$_}):('isnull'))) }
            grep { exists $params{$_} }
            keys %params ]);
    }

    # Make sure we have exactly one row returned...
    if(not @res)
    {
        carp $self->E_NOROWSFOUND;
        $self->__do_rollback;
        return;
    }
    if(@res > 1)
    {
        carp $self->E_TOOMANYROWS;
        $self->__do_rollback;
        return;
    }

    # Turn off warnings and commit
    local ($SIG{__WARN__}) = $self->w__noop;
    $handle->commit;
    return $res[0];
}

# Delete stuff from the database
sub delete
{
    my ($self, @criteria) = @_;
    my $handle = $self->handle;
    carp "can't delete without a database handle" and return unless $handle;
    carp "can't delete from a JOIN" and return if $self->_join_clause;

    # Create the SQL command
    my ($where, @binds) = $self->__create_where(@criteria);
    my $sql = "DELETE FROM " . $self->_sql_name;
    $sql .= " WHERE $where" if $where;

    unshift @binds, undef if @binds;    # Handle DBI lossage
    my $res = $handle->do($sql, @binds);
    
    # Report errors
    if(not $res)
    {
        carp "Database error " . $handle->errstr;
        $self->__do_rollback;
        return;
    }

    # Commit and return
    local ($SIG{__WARN__}) = $self->w__noop;
    $handle->commit;
    return $self;
}

# Search the database, return a row object per returned item
sub search
{
    my ($self, @criteria) = @_;
    my $handle = $self->handle;
    carp "can't search without a database handle" and return unless $handle;

    # create the base select statement
    my $sql = 'SELECT ' . join(',', $self->_column_sql_names) . ' FROM '
        . $self->_sql_name;

    # Add any join clause
    my $join = $self->_join_clause;
    $sql .= " $join" if $join;

    # Add a where clause, if necessary
    my ($where, @binds) = $self->__create_where(@criteria);
    $sql .= " WHERE $where" if $where;

    # Add any GROUP BY clause
    my @g = $self->_group_by;
    $sql .= " GROUP BY " . join(',',@g) if @g;

    # Add any ORDER BY clause
    my $ord = $self->__find_orderby(@criteria);
    $sql .= " $ord" if $ord;

    # Add any LIMIT clause
    my $limit = $self->__find_limit(@criteria);
    $sql .= " $limit" if $limit;
    
    unshift @binds, undef if @binds;
    my $data = $handle->selectall_arrayref($sql, @binds);

    carp "Database error " . $handle->errstr and return unless $data;

    # The return values row class
    my $rclass = $self->_row_class;
    $rclass = ref $self if $self->isa($rclass);

    # Create the return values
    my @res;
    for my $row (@$data)
    {
        my $robj = bless $self->new, $rclass;
        $robj->__set_data(@$row);

        # Add the where clause, so we can find this row later
        $robj->__create_where;
        push @res, $robj;
    }
    return @res;
}

# Return the number of rows a query would return
sub size
{
    my ($self, @criteria) = @_;
    my $handle = $self->handle;
    carp "can't find table size without a database handle" and return
        unless $handle;

    # Create the base SQL statement
    my $table = $self->_sql_name;
    my $sql = "SELECT COUNT(*) FROM $table";

    # Add any GROUP BY clause
    my @g = $self->_group_by;
    $sql .= " GROUP BY " . join(',',@g) if @g;

    # Add any join clause
    my $join = $self->_join_clause;
    $sql .= " $join" if $join;

    my ($where, @binds) = $self->__create_where(@criteria);
    $sql .= " WHERE $where" if $where;
    
    unshift @binds, undef if @binds;    # Avoid DBI lossage
    my $data = $handle->selectall_arrayref($sql, @binds);

    carp "Database error " . $handle->errstr and return unless $data;
    return $data->[0][0];
}

1;

__END__


lib/DBIx/ORM/Declarative.pm
package DBIx::ORM::Declarative;

use strict;
use Carp;
use DBIx::ORM::Declarative::Schema;
use DBIx::ORM::Declarative::Table;
use DBIx::ORM::Declarative::Join;
use DBIx::ORM::Declarative::Row;
use DBIx::ORM::Declarative::JRow;

use vars qw($VERSION);
$VERSION = '0.21';

use constant BASE_CLASS   => 'DBIx::ORM::Declarative';
use constant SCHEMA_CLASS => 'DBIx::ORM::Declarative::Schema';
use constant TABLE_CLASS  => 'DBIx::ORM::Declarative::Table';
use constant JOIN_CLASS   => 'DBIx::ORM::Declarative::Join';
use constant ROW_CLASS    => 'DBIx::ORM::Declarative::Row';
use constant JROW_CLASS   => 'DBIx::ORM::Declarative::JRow';

# Use this to /really/ supress warnings
use constant w__noop => sub { };

# The error we return when we have an embarassment of riches
use constant E_TOOMANYROWS => 'Database error: underdetermined data set';

# The error we return when we've lost the row we just inserted
use constant E_NOROWSFOUND => 'Database error: inserted data not found';

# We need to register table & join creation methods; otherwise, we
# may wind up blowing up when we try to deal with a table that has
# a column with the same name as the table itself
my %table_methods = ();
my %join_methods = ();

sub table_method
{
    my ($self, $table, $method) = @_;
    return $table_methods{$table} if $table_methods{$table} and not $method;
    $table_methods{$table} = $method;
}

sub join_method
{
    my ($self, $join, $method) = @_;
    return $join_methods{$join} if $join_methods{$join} and not $method;
    $join_methods{$join} = $method;
}

# This applies a method by name - necessary for perl < 5.6
sub apply_method
{
    my ($obj, $method, $wantarray, @args) = @_;
    # Check to see if we can apply it directly
    my @rv;
    eval
    {
        # We don't need any warnings here
        local $SIG{__WARN__} = __PACKAGE__->w__noop;
        if($wantarray)
        {
            @rv = $obj->$method(@args);
        }
        else
        {
            $rv[0] = $obj->$method(@args);
        }
    } ;
    if(not $@)
    {
        return $wantarray?@rv:$rv[0];
    }
    my $res = UNIVERSAL::can($obj, $method);
    if($res)
    {
        return $wantarray?($res->($obj, @args)):scalar($res->($obj, @args));
    }
    $res = UNIVERSAL::can($obj, 'AUTOLOAD');
    if($res)
    {
        # We can't directly use the result in $res, because we need to know
        # which AUTOLOAD it found.  Just use eval for now.  *sigh*.
        if($wantarray)
        {
            eval "\@rv = \$obj->$method(\@args)";
        }
        else
        {
            eval "\$rv[0] = \$obj->$method(\@args)";
        }
        carp $@ if $@;
        return $wantarray?@rv:$rv[0];
    }
    my $class = ref $obj || $obj;
    carp qq(Can't locate object method "$method" via package "$class");
}

# Create a new DBIx::ORM::Declarative object
# Accepts args as a hash
# Recognized args are "handle" and "debug"
# Unrecognized args are ignored
# If used as an object method, copy the handle and debug status, if available
sub new
{
    my ($self, %args) = @_;
    my $class = ref $self || $self;
    my $handle = exists $args{handle}?$args{handle}:$self->handle;
    my $debug = delete $args{debug} || $self->debug_level || 0;
    if(not exists $args{handle} and DBI->can('connect') and $args{dsn})
    {
        $handle = DBI->connect(@args{qw(dsn username password)},
            { RaiseError => 0, PrintError => 0, AutoCommit => 0 });
    }
    my $rv = bless { __handle => $handle, __debug_level => $debug }, $class;
    return $rv;
}

# Custom import method to create schemas during the "use" clause.
sub import
{
    my ($package, @args) = @_;
    if(not ref $args[0])
    {
        $package->schema(@args);
        return;
    }
    for my $arg (@args)
    {
        if(not ref $arg)
        {
            carp "Can't import '$arg' in '$package'";
            next;
        }
        $package->schema(%$arg);
    }
}

# Get or set the DBI handle
sub handle
{
    my $self = shift;
    return unless ref $self;
    if(@_)
    {
        delete $self->{__handle};
        $self->{__handle} = $_[0] if $_[0];
        return $self;
    }
    return unless exists $self->{__handle};
    return $self->{__handle};
}

# Get or set the debug level
sub debug_level
{
    my $self = shift;
    return 0 unless ref $self;
    if(@_)
    {
        $self->{__debug_level} = $_[0] || 0;
        return $self;
    }
    return $self->{__debug_level} || 0;
}

# Get the current schema name, or switch to a new schema, or create a
# new schema class.
sub schema
{
    my ($self, @args) = @_;
    if(@args<2)
    {
        if(@args==1)
        {
            my $schema = shift @args;
            return $self->apply_method($schema,wantarray)
                if $schema and $self->can($schema);
            return $self;
        }
        my $schema;
        eval { $schema = $self->_schema; };
        return $schema;
    }

    # Creating/updating a schema class - process the args
    my %args = @args;
    my $schema = delete $args{schema};
    my $from_dual = delete $args{from_dual};
    my $limit_clause = delete $args{limit_clause} || 'LIMIT %offset%,%count%';

    carp "missing schema argument" and return unless $schema;
    my $schema_class = $self->SCHEMA_CLASS . "::$schema";

    # The meat of the declarations
    my $tables = delete $args{tables} || [ ];
    my $joins = delete $args{joins} || [ ];
    my $aliases = delete $args{table_aliases} || { };

    # We're gonna do a whole mess of symbolic references...
    no strict 'refs';
    my $schema_method_name = $self->BASE_CLASS . "::$schema";
    if(not @{$schema_class . '::ISA'})
    {
        # Create the class heirarchy
        @{$schema_class . '::ISA'} = ($self->SCHEMA_CLASS);

        # Let's see if we're called from import...
        my ($pkg, $file, $line, $sub) = caller(1);
        if($sub eq __PACKAGE__ . '::import' and $pkg ne 'main')
        {
            # Yep - insert ourselves in the upstream @ISA...
            my $isaref = \@{$pkg . '::ISA'};
            push @$isaref, $schema_class
                unless grep { $_ eq $schema_class } @$isaref;
        }

        # Information methods
        *{$schema_class . '::_schema' } = sub { $schema; };
        *{$schema_class . '::_schema_class' } =
        *{$schema_class . '::_class' } = sub { $schema_class; };
        *{$schema_class . '::_limit_clause' } = sub { $limit_clause; };
        *{$schema_class . '::_from_dual' } = sub { $from_dual; };

        # A constructor/mutator
        *{$schema_method_name} = sub
        {
            my ($self) = @_;
            my $rv = $self->new;
            bless $rv, $schema_class unless $rv->isa($schema_class);
            return $rv;
        } ;
    }

    # Create the tables
    $schema_class->table(%$_) foreach @$tables;

    # Create the aliases, if we have any
    $schema_class->alias($_, $aliases->{$_}) foreach keys %$aliases;
    
    # Create any joins we might have
    $schema_class->join(%$_) foreach @$joins;

    return &{$schema_method_name}($self);
}

1;
